include/rdma/fabric.h

```diff
@@ -523,6 +524,35 @@ struct fi_wait_attr;
 /* fi_bind()-specific flags */
 #define FI_SELECTIVE_COMPLETION	(1ULL << 59)
 
+enum fi_hmem_iface {
+	FI_HMEM_SYSTEM	= 0,
+	FI_HMEM_CUDA,
+};
+
+#define FI_SET_OPS_HMEM_OVERRIDE "hmem_override_ops"
+
+struct fi_hmem_override_ops {
+	size_t	size;
+
+	/* Copy from HMEM iov to host buffer. Number of bytes copied must be
+	 * returned on success. Negative fi_errno must be returned on error.
+	 */
+	ssize_t	(*copy_from_hmem_iov)(void *dest, size_t size,
+				      const struct iovec *hmem_iov,
+				      enum fi_hmem_iface hmem_iface,
+				      size_t hmem_iov_count,
+				      uint64_t hmem_iov_offset);
+
+	/* Copy to HMEM iov from host buffer. Number of bytes copied must be
+	 * returned on success. Negative fi_errno must be returned on error.
+	 */
+	ssize_t (*copy_to_hmem_iov)(const struct iovec *hmem_iov,
+				    enum fi_hmem_iface hmem_iface,
+				    size_t hmem_iov_count,
+				    uint64_t hmem_iov_offset, void *src,
+				    size_t size);
+};
```

Comment:
This define may be a better fit in fi_domain.h.  Though, we should decide how we want to handle get/set ops.  For example:  Do we define each in the best existing header file?  Do we create a new header file and put most ops there?  Do we handle these on a case by case basis?

I *think* these ops should be set per domain, since that's the scope of memory registration.

Reply:
If we move this into fi_domain.h, I think this may set a precedence where future overrides are associated with a specific fabric object (domain, ep, etc...). This seems to point to overrides being defined within the associated object header file. In addition, seems like each object would have to define a set operation.

My motivation for having the set operation associated with a fid was to allow for a single set operation interface to be defined. It would be up to the object containing the fid to accept/reject the set operation call. With this approach, I think it would make sense to have all overrides define in a single location.

> I think these ops should be set per domain, since that's the scope of memory registration.

While memory registration is at the domain level, it is really the endpoints which would be preforming the copy routine. So if we do associate these ops with a domain, should all endpoints inherit ops from the domain?

Reply:
There are separate issues here.

Where should a setops call go?  The current placement looks correct.  This should be generic.

Which object(s) support the hmem override?  I think this makes sense to associate memory region (registration) handling.  That implies that regardless of which endpoint a region is accessed through, the behavior is the same.  This also reduces the per endpoint overhead.  Associating this with the domain, versus a MR, handles the case where the provider does not require explicit memory registration.

I expect future overrides to be associated with a well defined set of fabric objects.

Reply:
> Where should a setops call go? The current placement looks correct. This should be generic.

Sounds good.

> Which object(s) support the hmem override? I think this makes sense to associate memory region (registration) handling. That implies that regardless of which endpoint a region is accessed through, the behavior is the same. This also reduces the per endpoint overhead. Associating this with the domain, versus a MR, handles the case where the provider does not require explicit memory registration.

Works for me.

> I expect future overrides to be associated with a well defined set of fabric objects.

Ok. Seems to make sense to store the override definitions in the associated fabric object header. For this case, I can move FI_SET_OPS_HMEM_OVERRIDE into fi_domain.h. Sound good?

Reply:
Yes - sounds reasonable